README.md
# blocklift-js

A dead simple and developer friendly JavaScript library for handling object storage on Azure

[![Build Status](https://travis-ci.org/julie-ng/blocklift-js.svg?branch=master)](https://travis-ci.org/julie-ng/blocklift-js)
[![Test Coverage](https://api.codeclimate.com/v1/badges/4b5bf0e191de7c6c74c8/test_coverage)](https://codeclimate.com/github/julie-ng/blocklift-js/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/4b5bf0e191de7c6c74c8/maintainability)](https://codeclimate.com/github/julie-ng/blocklift-js/maintainability)

## Developer Friendly API

For a simple blob operation, the official [@azure/storage-blob](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/storage/storage-blob) SDK requires 3 "clients" for the blob service, the container and the blob itself. Blocklift.js abstracts away the complexity to provide a better developer experience.

This is a complete example to upload a file in just a few lines:

```javascript
const Blocklift = require('blocklift')

const lift = new Blocklift({
	account: 'accountname',
	accessKey: 'key',
	defaultContainer: 'dev' // optional
})

lift.upload('hello.txt', 'Hello World')
	.then((upload) => { console.log(upload.url) })
	.catch((err) => { … })
```

You can also use the `async/await` syntax:

```javascript
async function main () {
	const blob =  await lift.upload('hello.txt', 'Hello World')
	console.log(blob.url)
}
main()
```

### Easier than the official SDK

Unfortunately the official [@azure/storage-blob](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/storage/storage-blob) SDK is cumbersome to use. 

Compare the above with this [official example](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/storage/storage-blob#create-a-blob-by-uploading-data-to), which has been slightly simplified for fairer comparison:

```javascript
const { DefaultAzureCredential } = require("@azure/identity");
const { BlobServiceClient } = require("@azure/storage-blob");
const defaultAzureCredential = new DefaultAzureCredential();
const account = ACCOUNT_NAME;

const blobServiceClient = new BlobServiceClient(
  `https://${ACCOUNT_NAME}.blob.core.windows.net`,
  defaultAzureCredential
);

const content = "Hello world!";
const containerClient = blobServiceClient.getContainerClient('dev');
const blockBlobClient = containerClient.getBlockBlobClient('hello.txt');
blockBlobClient.upload(content, content.length)
	.then((response) => { … })
	.catch((err) => { … })
```

### Automatically get Blob URLs

In addition to standard Blob REST API response, blocklift.js also returns a `url` property for you, for example:

```
https://accountname.blob.core.windows.net/dev/hello.txt
```

### Set a default container

Why repeatedly pass a container parameter or client, when you can just set a default once and then forget about it?

```javascript
const lift = new Blocklift({
	account: ACCOUNT_NAME,
	accessKey: ACCESS_KEY,
	defaultContainer: 'dev' // must already exist
})
```

Of course, you can always choose to use a different container and pass an _optional_ `container` parameter in individual operations:

```javascript
const options = { conatiner: 'not-default' }
lift.upload('hello.txt', 'Hello World', options)
```

### Automatic `Content-Type` detection

Blocklift.js automatically sets a `Content-Type` based on file contents or filename. This helps browsers identify file types and decide how to handle them. You want to show users content as quickly as possible, i.e. in browser and not have them dig through their Downloads folder.

| Content Type | Browser Behavior |
|:--|:--|
| none | browser downloads file (default Azure SDK behavior) |
| `text/plain` | display text in browser |
| `image/jpg` | display image in browser |

Of course other file types are supported including PDFs, Microsoft Word, Powerpoint, Videos, and more. See [file-type](https://www.npmjs.com/package/file-type#supported-file-types) package for detailed list.

## Examples

Everything is a `Promise`. For a better overview, the `then()`, `catch()`, `async`, and `await` have been left out.

For full details, see [the API documentation &rarr;](https://julie-ng.github.io/blocklift-js/)

### Containers

```javascript
lift.listContainers()
lift.createContainer('name')
lift.deleteContainer('name')
```

### Blobs

```javascript
lift.listBlobs('container')
lift.getBlobUrl('my-image.png')
```

Upload a Blob or Create a File

```javascript
lift.upload('hello.txt', 'Hello World', { contentType: 'text/plain' })
lift.uploadFile('local-file.png', 'folder/image.png')
```

Delete a Blob

```javascript
lift.deleteBlob('folder/image.png')
lift.deleteBlob('folder/image.png', { container: 'not-default' })
```

### Responses

Blocklift.js uses the official SDK under the hood and will bubble up the responses. 

For example, `url` is added by blocklift and `serverResponse` is the unaltered response from the SDK.

```
{
  url: 'https://myaccount.blob.core.windows.net/default/hello-2020-03-22.txt',
  serverReponse: {
    etag: '"0x8D7CE6FCC47FC45"',
    lastModified: 2020-03-22T14:46:26.000Z,
    contentMD5: <Buffer c4 85 73 58 38 b0 05 75 7f 3f 31 99 50 0c 0e ca>,
    clientRequestId: 'd749b9a2-5fba-489d-a230-ba5df662f4f0',
    requestId: '283595f3-a01e-0009-1b58-00e0fc000000',
    version: '2019-02-02',
    date: 2020-03-22T14:46:25.000Z,
    isServerEncrypted: true,
    encryptionKeySha256: undefined,
    errorCode: undefined,
    'content-length': '0',
    server: 'Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0',
    'x-ms-content-crc64': 'awyymjQimGI=',
    body: undefined
  }
}
```

For more examples, see [the API documentation &rarr;](https://julie-ng.github.io/blocklift-js/Blocklift.html)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/julie-ng/blocklift-js/compare/v1.0.0...v1.0.1) (2020-09-14)

## 1.0.0 (2020-03-25)


### Features

* **api:** rename deleteFile to deleteBlob ([b2c976c](https://github.com/julie-ng/blocklift-js/commit/b2c976c12eada1f288cf3bb8b8625b9ccf620602))
* **delete:** file ([01e9242](https://github.com/julie-ng/blocklift-js/commit/01e92426f6e5fe44ead050b558fa0b0c93367b19))
* **docs:** add autogenerated API docs with jsdoc ([c8aa502](https://github.com/julie-ng/blocklift-js/commit/c8aa50238c2fb45b5aab2de4205d7f8148942267))
* **upload:** file and content, and set content type header ([7bcbee1](https://github.com/julie-ng/blocklift-js/commit/7bcbee109865bdd5e549044c063a49266f41cc3f))


source/blocklift.js
const AzureSDK = require('./azure-sdk')
const ClientError = require('./client-error')
const util = require('./util')

/**
 * ## Syntax of your choice: `async`/`await` or `Promise`
 *
 * Blocklift.js is designed dead simple and will accept either syntax of your choice.
 * Unlike the Azure SDK, Blocklift doesn't care and is bilingual.
 * Just make sure to use and catch errors appropriately.
 *
 * **NOTE** - examples in documentation below do not include either syntax to avoid repetition.
 * It is your responsibility to implement with either Promises or async/await syntax.
 *
 * #### Use with classic `Promise` syntax
 *
 * ```javascript
 * lift.createContainer('name')
 *   .then(res => console.log(res))
 *   .catch(err => console.log(err))
 * ```
 *
 * #### Use with `async`/`await` syntax
 *
 * ```javascript
 * async main () {
 *   try {
 *     const res = await lift.createContainer('name')
 *     console.log(res)
 *   }
 *   catch ((err) => {
 *     console.log(err)
 *   })
 * }
 *
 * main()
 * ```
 *
 * ## Responses
 *
 * The Azure SDK leverages the REST API and server responses are generally passed through to Blocklift. In some cases, the responses are extended to include some handy properities, e.g. URL of uploaded blob.
 *
 *
 */
class Blocklift {

	/**
	 * Creates a new Blocklift client for Azure Storage for blobs (binary large objects).
	 *
	 * @param {String} opts.account - Blob Storage Account name
	 * @param {String} opts.accessKey - Blob Storage Account Access Key
 	 * @example
   * const lift = new Blocklift({
   *   account: 'your account name',
   *   accessKey: 'your access key',
   *   defaultContainer: 'your-default-container' // must already exist
   * })
	 */
	constructor (opts = {}) {

		/**
		 * [Azure Storage Account name](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview#naming-storage-accounts) e.g. `mystorageaccount`
		 *
		 * @readonly
		 * @type {String}
		 */
		this.account = opts.account

		/**
		 * Hostname for Azure Storage Account e.g. `https://myaccount.blob.core.windows.net`
		 *
		 * @private
		 * @readonly
		 * @type {String}
		 */
		this.host = `https://${opts.account}.blob.core.windows.net`

		/**
		 * Name of default container for blobs, used if container is not specified in API calls.
		 *
		 * @readonly
		 * @type {String|Object}
		 */
		this.defaultContainer = opts.defaultContainer || null

		/**
		 * Reference to AzureSDK Wrapper
		 *
		 * @private
		 * @readonly
		 * @type {AzureSDK}
		 */
		this.sdk = new AzureSDK(opts)
	}

	// -------- Containers --------

	/**
	 * Creates a new container
	 *
	 * Will throw an error if name is not all lowercase.
	 *
	 * @param {String} name - will be transformed to lowercase per Azure requirements
	 * @returns {Promise}
	 * @example
	 * lift.createContainer('name')
	 *
	 */
	createContainer (name) {
		if (name.toLowerCase() !== name) {
			throw `Error: invalid container name. Please use all lowercase letters`
		}

		return this.sdk.createContainer(name)
			.then((res) => {
				return {
					name: name,
					operation: 'create',
					message: 'Successfully created container ' + name,
					serverResponse: res
				}
			})
			.catch((err) => { throw new ClientError(err) })
	}

	/**
	 * Delete a container
	 *
	 * @param {String} name - container name
	 * @returns {Promise}
	 * @example
	 * lift.deleteContainer('name')
	 *
	 */
	deleteContainer (name) {
		return this.sdk.deleteContainer(name)
			.then((res) => {
				return {
					name: name,
					operation: 'delete',
					message: 'Successfully deleted container ' + name,
					serverResponse: res
				}
			})
			.catch((err) => { throw new ClientError(err) })
	}


	/**
	 * List Containers
	 *
	 * @returns {Promise<Array>} list of containers
	 * @returns {Promise<Object>} Error
	 * @example
	 * lift.listContainers()
	 *
	 * @example <caption>Example Response</caption>
   * [
   *   {
   *     name: 'container-1',
   *     properties: {…},
   *     metadata: undefined
   *   },
   *   {
   *     name: 'container-2',
   *     properties: {…},
   *     metadata: undefined
   *   }
   * ]
	 */
	listContainers () {
		return this.sdk.listContainers()
			.then((res) => res)
			.catch((err) => { throw new ClientError(err) })
	}

	// -------- Blobs --------

	/**
	 * List Blobs in a container or `defaultContainer` if none is provided.
	 *
	 * @param {String} [containerName]
	 * @returns {Promise<Array>} list of blobs
	 * @return {Promise<Object>} Error
	 * @example
	 * // list blobs in default container
	 * lift.listBlobs()
	 *
	 * // list blobs in specific container
	 * lift.listBlobs('container')
	 */
	listBlobs (containerName) {
		return this.sdk.listBlobs(containerName)
			.then((res) => res)
			.catch((err) => { throw new ClientError(err) })
	}

	/**
	 * Upload content
	 * at the moment not recommended because it does not include headers
	 * for example `Content-Type`
	 *
	 * @param {String} pathname
	 * @param {String} content
	 * @param {*} [opts={}]
	 * @returns {Promise}
	 * @example
	 * lift.upload('hello.txt', 'Hello World', { contentType: 'text/plain' })
	 *
	 * @example <caption>Example reponse</caption>
	 * {
   *   url: 'https://myaccount.blob.core.windows.net/default/hello-2020-03-22.txt',
   *   serverReponse: {
   *     etag: '"0x8D7CE6FCC47FC45"',
   *     lastModified: 2020-03-22T14:46:26.000Z,
   *     contentMD5: <Buffer c4 85 73 58 38 b0 05 75 7f 3f 31 99 50 0c 0e ca>,
   *     clientRequestId: 'd749b9a2-5fba-489d-a230-ba5df662f4f0',
   *     requestId: '283595f3-a01e-0009-1b58-00e0fc000000',
   *     version: '2019-02-02',
   *     date: 2020-03-22T14:46:25.000Z,
   *     isServerEncrypted: true,
   *     encryptionKeySha256: undefined,
   *     errorCode: undefined,
   *     'content-length': '0',
   *     server: 'Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0',
   *     'x-ms-content-crc64': 'awyymjQimGI=',
   *     body: undefined
   *   }
   * }
	 */
	async upload (pathname, content, opts = {}) {
		if (!pathname) {
			throw 'Error: missing required Blob `pathname` parameter, e.g. `foo.txt` or `foo/bar.txt`'
		}
		const container = opts.container || this.defaultContainer
		const requiredParams = {
			container: container,
			pathname: pathname,
			content: content
		}

		return this.sdk.upload(requiredParams, opts)
			.then((res) => {
				return {
					url: this.getBlobUrl(pathname, container),
					serverReponse: res
				}
			})
			.catch((err) => {
				throw new ClientError(err)
			})
	}

	/**
	 * Upload a file. Blocklift will attempt to set the `Content-Type` for you
	 * based on file extension and/or contents.
	 *
	 * @param {String} sourcePath - source file pathname
	 * @param {String} blobPath - desintation file pathname on Azure
	 * @param {Object} [opts = {}]
	 * @returns {Promise}
	 * @example
	 * lift.uploadFile('local-file.png', 'folder/image.png')
	 *
	 * @example <caption>Example reponse</caption>
	 * {
   *   url: 'https://myaccount.blob.core.windows.net/default/hello-2020-03-22.txt',
   *   serverReponse: {…}
   * }
	 */
	async uploadFile (sourcePath, blobPath, opts = {}) {
		const container = opts.container || this.defaultContainer
		const contentType = await util.getContentTypeFromFile(sourcePath)
		const requiredParams = {
			container: container,
			filePathname: sourcePath,
			blobPathname: blobPath,
		}
		opts = {
			...opts,
			contentType: contentType
		}

		return this.sdk.uploadFile(requiredParams, opts)
			.then((res) => {
				return {
					url: this.getBlobUrl(blobPath, container),
					serverReponse: res
				}
			})
			.catch((err) => {
				throw new ClientError(err)
			})
	}

	/**
	 * Deletes a Blob from the Storage Account
	 *
	 * @param {String} pathname path to file
	 * @param {Object} [opts]
	 * @param {String} [opts.container] file's container name
	 * @returns {Promise}
	 * @example
	 * lift.deleteBlob('folder/image.png') // uses default container
	 * lift.deleteBlob('folder/image.png', { container: 'my-container' })
	 *
	 * @example <caption>Example Response</caption>
	 * {
   *   name: 'container-2020-03-22-1519-a7uotq',
   *   operation: 'delete',
   *   message: 'Successfully deleted container container-2020-03-22-1519-a7uotq',
   *   serverResponse: {
   *     clientRequestId: '0cf51581-ebbe-4b4c-aaca-4d31bbad1b58',
   *     requestId: '0a6759b4-a01e-0054-0154-00ea78000000',
   *     version: '2019-02-02',
   *     date: 2020-03-22T14:19:55.000Z,
   *     errorCode: undefined,
   *     'content-length': '0',
   *     server: 'Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0',
   *     body: undefined
   *   }
   * }
	 */
	async deleteBlob (pathname, opts = {}) {
		const params = {
			container: opts.container || this.defaultContainer,
			pathname: pathname
		}
		return this.sdk.deleteBlob(params)
			.then((res) => res)
			.catch((err) => { throw new ClientError(err) })
	}

	/**
	 * Helper that returns URL for a given blob in a given container, e.g. `https://myaccount.blob.core.windows.net/default/my-image.png`
	 *
	 * @param {String} pathname path to file
	 * @param {String} [container] container name
	 * @return {String}
	 * @example
	 * lift.getBlobUrl('my-image.png')
	 *
	 */
	getBlobUrl (pathname, container = '') {
		container = (container === '')
			? this.defaultContainer
			: container
		return this.host + `/${container}/${pathname}`
	}
}


module.exports = Blocklift
